src/providers/aws.rs

```diff
@@ -0,0 +1,887 @@
+    ///
+    /// The default region is us-east-1. [Available regions are listed
+    /// here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-available-regions)
+    ///
+    /// AMIs are region-specific. This will overwrite the ami field to
+    /// the Ubuntu 18.04 LTS AMI in the selected region.
+    pub fn region_with_ubuntu_ami(mut self, region: Region) -> Setup<YesAmi> {
+        self.region = region.clone();
+        let ami: String = UbuntuAmi::from(region).into();
+        self.ami(ami)
+    }
+
+    /// The new instance will start out in the state dictated by the Amazon Machine Image specified
+    /// in `ami`. Default is Ubuntu 18.04 LTS.
+    pub fn ami(self, ami: impl ToString) -> Setup<YesAmi> {
+        Setup {
+            region: self.region,
+            instance_type: self.instance_type,
+            ami: Some(ami.to_string()),
+            setup: self.setup,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+
+    /// The given AWS EC2 instance type will be used. Note that only [EC2 Defined Duration Spot
+    /// Instance types](https://aws.amazon.com/ec2/spot/pricing/) are allowed.
+    pub fn instance_type(mut self, typ: impl ToString) -> Self {
+        self.instance_type = typ.to_string();
+        self
+    }
+
+    /// The provided callback, `setup`, is called once for every spawned instances of this type with a handle
+    /// to the target machine. Use [`Machine::ssh`] to issue
+    /// commands on the host in question.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tsunami::providers::aws::Setup;
+    ///
+    /// let m = Setup::default()
+    ///     .setup(|ssh, log| {
+    ///         slog::info!(log, "running setup!");
+    ///         ssh.cmd("sudo apt update")?;
+    ///         Ok(())
+    ///     });
+    /// ```
+    pub fn setup(
+        mut self,
+        setup: impl Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync + 'static,
+    ) -> Self {
+        self.setup = Some(Arc::new(setup));
+        self
+    }
+}
+
+impl Setup<YesAmi> {
+    /// Set up the machine in a specific EC2
+    /// [`Region`](http://rusoto.github.io/rusoto/rusoto_core/region/enum.Region.html).
+    ///
+    /// The default region is us-east-1. [Available regions are listed
+    /// here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-available-regions)
+    ///
+    /// AMIs are region-specific.
+    /// This will clear the AMI field, which must be set for this struct to be useful.
+    pub fn region(self, region: Region) -> Setup<NoAmi> {
+        Setup {
+            region,
+            instance_type: self.instance_type,
+            ami: None,
+            setup: self.setup,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+/// AWS EC2 spot instance launcher.
```

Comment:
It'd probably be good to point people towards `TsunamiBuilder::spawn` here, since that's how they're likely to _use_ this type.